Add Global Auditing Disable Feature

snopoke opened the pull request:
> [!NOTE]
> This code was originally written by Claude AI
 
## Summary

Adds the ability to globally disable auditing in django-field-audit, providing both application-wide control via Django settings and fine-grained runtime control via context managers. This feature is particularly useful for unit tests, data migrations, bulk imports, and maintenance operations where audit overhead is not needed.

## Implementation

### Core Infrastructure

- **Context Variable**: Uses Python's `contextvars` for thread-safe, async-safe state management
- **Check Function**: `is_audit_enabled()` checks context variable first, then falls back to Django setting
- **Context Managers**: `disable_audit()` and `enable_audit()` for runtime control

### API

New public API exports:
```python
from field_audit import disable_audit, enable_audit
```

New Django setting:
```python
FIELD_AUDIT_ENABLED = False  # Default: True
```

## Usage Examples

### Global Disable via Setting

```python
# settings.py
FIELD_AUDIT_ENABLED = False
```

### Runtime Disable

```python
from field_audit import disable_audit

# Temporarily disable auditing
with disable_audit():
    obj.save()  # No audit event created
    MyModel.objects.bulk_create(objects)  # No audit events
```

### Override Global Setting

```python
from field_audit import enable_audit

# When FIELD_AUDIT_ENABLED=False, temporarily enable
with enable_audit():
    obj.save()  # Audit event IS created
```

### Use Cases

**Unit Tests** - Improve performance by skipping audit overhead:
```python
def test_without_audit(self):
    with disable_audit():
        obj = MyModel.objects.create(field="test")
```

**Data Migrations** - Skip auditing during bulk operations:
```python
def migrate_data():
    with disable_audit():
        MyModel.objects.filter(old=True).update(new=True)
```

### Backwards Compatibility

The new features added in this PR are opt-in and do not break backward compatibility.

snopoke committed:
Add core infrastructure for global audit disable

- Add audit_enabled context variable
- Add is_audit_enabled() check function
- Add disable_audit() and enable_audit() context managers
- Update __all__ exports

snopoke committed:
Add audit disable checks to model operations

- Add check to _decorate_db_write for save/delete
- Add check to _m2m_changed_handler for M2M changes

snopoke committed:
Add audit disable checks to service methods

- Add check to audit_field_changes
- Add check to create_audit_event

snopoke committed:
Add audit disable checks to QuerySet methods

- Add check to bulk_create
- Add check to delete
- Add check to update

snopoke committed:
Export disable_audit and enable_audit from package

snopoke committed:
Add comprehensive test suite for global disable

- Test setting-based disable
- Test context manager disable/enable
- Test M2M field auditing
- Test QuerySet operations
- Test thread safety
- Test backwards compatibility
- Test migration scenarios

snopoke committed:
Fix test suite and simplify thread safety test

- Update global disable tests to use correct model names
- Fix test isolation issues
- Simplify thread safety test to avoid Django transaction complexity
- All 200 tests now pass

snopoke committed:
Add documentation for global audit disable feature

- Add FIELD_AUDIT_ENABLED to settings table
- Add Disabling Auditing section with examples
- Document disable_audit() and enable_audit() context managers
- Include use cases for tests, migrations, and imports

snopoke committed:
cleanup tests

millerdev commented on GLOBAL_DISABLE_PLAN.md:
For my curiosity, how carefully did you read all 1000+ lines of this plan doc? Did you write all of it, or was it mostly AI-generated?

Should this file be kept after the plan is executed? Seems like the root of the repository may become quite cluttered if many of these were created. Maybe move it somewhere out of the way if it should be kept?

snopoke replied:
This wasn't meant to be kept in the repo. I did read it quite carefully.

millerdev commented on field_audit/field_audit.py:
```diff
@@ -9,12 +9,81 @@
+    Returns True if auditing should proceed, False if disabled.
+    Checks context variable first, then falls back to Django setting.
+    """
+    from django.conf import settings
```
Why is this a local import?

The same applies to other imports below.

millerdev commented on field_audit/field_audit.py:
```diff
@@ -9,12 +9,81 @@
+    from contextlib import contextmanager
+
+    @contextmanager
+    def _context():
```
Why the inner function? Why not...
```py
@contextmanager
def disable_audit():
    ...
```

millerdev commented on field_audit/models.py:
```diff
@@ -732,6 +735,11 @@ def delete(self, *, audit_action=AuditAction.RAISE):
+
+        from .field_audit import is_audit_enabled
+        if not is_audit_enabled():
+            return super().delete()
```
Why not do this check above?
```suggestion
        if audit_action is AuditAction.IGNORE or not is_audit_enabled():
            return super().delete()
        assert audit_action is AuditAction.AUDIT, audit_action
```

These suggestions can be applied in all other changes in this file as well.

snopoke replied:
So that the assert for the AuditAction is still applied even if auditing is disabled. 

My logic was that if you disabled auditing in tests you'd still want to know if this validation failed.

But for other use cases I guess it might not make sense.

snopoke replied:
Looking again, the keyword validation is done in the decorator so it would be fine to combine these two.

millerdev commented on field_audit/models.py:
```diff
@@ -703,7 +703,10 @@ def bulk_create(self, objs, *, audit_action=AuditAction.RAISE, **kw):
         assert audit_action is AuditAction.AUDIT, audit_action
 
-        from .field_audit import request
+        from .field_audit import is_audit_enabled, request
```
What do you think of moving `is_audit_enabled` and related decorators to a different module (other than `field_audit/field_audit.py`) so they can be imported at the top of this file? I don't think they have any dependencies that would cause a circular import or otherwise require local import.

millerdev commented on tests/test_global_disable.py:
```diff
@@ -0,0 +1,276 @@
+    """Test FIELD_AUDIT_ENABLED setting."""
+
+    @override_settings(FIELD_AUDIT_ENABLED=False)
+    def test_save_disabled_via_setting(self):
```
Should this test clear or maybe assert that there is no thread-local state? Seems like that might cause false positives. The same applies to other tests in this class.

millerdev commented on tests/test_global_disable.py:
```diff
@@ -0,0 +1,276 @@
+
+        self.assertEqual(AuditEvent.objects.count(), 0)
+
+    def test_save_enabled_by_default(self):
```
Noticing the absence of `test_delete_enabled_by_default`. Is this test even necessary since it's probably covered by other tests that verify save logic?

millerdev commented on tests/test_global_disable.py:
```diff
@@ -0,0 +1,218 @@
+        """Verify auditing re-enabled after exception in context."""
+        try:
+            with disable_audit():
+                SimpleModel.objects.create(value="test")
```
Is the database hit necessary?
```suggestion
```

millerdev commented on tests/test_global_disable.py:
```diff
@@ -0,0 +1,218 @@
+
+        # Auditing should be re-enabled
+        SimpleModel.objects.create(value="test2")
+        self.assertEqual(AuditEvent.objects.count(), 1)
```
DB hit necessary?

```suggestion
        assert is_audit_enabled()
```

millerdev commented on tests/test_global_disable.py:
```diff
@@ -0,0 +1,218 @@
+
+        # Enable for specific operation
+        with enable_audit():
+            obj2 = SimpleModel.objects.create(value="test2")
```
Database hits seem unnecessary in this test. Instead `assert is_audit_enabled()`

millerdev commented on tests/test_global_disable.py:
```diff
@@ -0,0 +1,218 @@
+            ModelWithAuditingManager.objects.create(value=f"test{i}")
+            for i in range(3)
+        ]
+        AuditEvent.objects.all().delete()  # Clear create events
```
Why not use `bulk_create(..., audit_action=AuditAction.IGNORE)` here?

millerdev commented on tests/test_global_disable.py:
```diff
@@ -0,0 +1,218 @@
+        """Verify QuerySet.delete respects global disable."""
+        for i in range(3):
+            ModelWithAuditingManager.objects.create(value=f"test{i}")
+        AuditEvent.objects.all().delete()
```
Simplify to `bulk_create(..., audit_action=AuditAction.IGNORE)`?

millerdev commented on tests/test_global_disable.py:
```diff
@@ -0,0 +1,218 @@
+
+
+class BackwardsCompatibilityTestCase(TestCase):
+    """Verify existing behavior unchanged when feature not used."""
```
Are these meaningful tests? Is it true that other tests that existed prior to this feature being developed already cover this?

snopoke replied:
Not strictly necessary but I did think it was nice to have them together with the others. But I've removed them now.

millerdev commented on tests/test_global_disable.py:
```diff
@@ -0,0 +1,218 @@
+
+
+class ContextDisableTestCase(TestCase):
+    """Test disable_audit() context manager."""
```
Do these tests need to hit the database? Could they use `is_audit_enabled()` instead?

Left a few more specific comments in a single test below, but I think the general rule applies to all tests in this class.

millerdev reviewed (commented):
> **Zero performance impact** when enabled: Check overhead <1μs

Is this always true? Is this AI-generated PR description content? Seems like the entire **Backwards Compatibility** section could be removed or replaced with a single line: "The new features added in this PR are opt-in and do not break backward compatibility."

---

This PR seems unnecessarily verbose—and I didn't even read the whole plan doc, which feels wrong because it's slated to be kept in the repo. Wading through AI slop is tiring. The time savings gained by using AI seem to have translated to extra review time. That's frustrating.

snopoke committed:
remove unnecessary tests

snopoke committed:
simplify tests to check state directly

snopoke committed:
remove test

snopoke committed:
check state before test

snopoke committed:
move imports to top level

snopoke committed:
remove inner methods

snopoke committed:
move utils to new package remove local imports

snopoke committed:
combine if statements

snopoke committed:
move imports to module level

snopoke commented:
@millerdev apologies for the rebase, I wasn't thinking.

SmittieC reviewed (approved):
Looks fine to me

millerdev reviewed (approved):
> @millerdev apologies for the rebase, I wasn't thinking.

Looks like the plan doc was somehow completely removed in the rebase, and surprisingly I don't even see it removed in the [before/after rebase compare view](https://github.com/dimagi/django-field-audit/compare/6e97e9c20b42552788eb0ad38e008843559648d8..4eb7d1fd87a3656bc61979527b916b30cb8b8319) (although it's in the [previous branch history](https://github.com/dimagi/django-field-audit/commit/6e97e9c20b42552788eb0ad38e008843559648d8)). I guess that's fine.

Maybe I missed it, but I think there is no attribution to AI code generation anywhere in this PR anymore. How do you think about that? Is it reasonable to not mention when AI is used to write code?

snopoke commented:
> > @millerdev apologies for the rebase, I wasn't thinking.
> 
> Looks like the plan doc was somehow completely removed in the rebase, and surprisingly I don't even see it removed in the [before/after rebase compare view](https://github.com/dimagi/django-field-audit/compare/6e97e9c20b42552788eb0ad38e008843559648d8..4eb7d1fd87a3656bc61979527b916b30cb8b8319) (although it's in the [previous branch history](https://github.com/dimagi/django-field-audit/commit/6e97e9c20b42552788eb0ad38e008843559648d8)). I guess that's fine.
> 
> Maybe I missed it, but I think there is no attribution to AI code generation anywhere in this PR anymore. How do you think about that? Is it reasonable to not mention when AI is used to write code?

I thought the original commits had co-authorship - but maybe not since I rebased.

I've added a note at the top of the PR description.
